python/runtime/cudaq/algorithms/py_sample_async.cpp

```diff
@@ -50,43 +53,68 @@ for more information on this programming pattern.)#")
         auto kernelMod = kernel.attr("module").cast<MlirModule>();
         args = simplifiedValidateInputArguments(args);
-        auto *argData = toOpaqueArgs(args, kernelMod, kernelName);
+
+        // This kernel may not have been registered to the quake registry
+        // (usually, the first invocation would register the kernel)
+        // i.e., `cudaq::kernelHasConditionalFeedback` won't be able to tell if
+        // this kernel has qubit measurement feedback on the first invocation.
+        if (cudaq::get_quake_by_name(kernelName, false, std::nullopt).empty()) {
+          auto moduleOp = unwrap(kernelMod);
+          std::string mlirCode;
+          llvm::raw_string_ostream outStr(mlirCode);
+          mlir::OpPrintingFlags opf;
+          moduleOp.print(outStr, opf);
+          cudaq::registry::__cudaq_deviceCodeHolderAdd(kernelName.c_str(),
+                                                       mlirCode.c_str());
+        }
 
         // The function below will be executed multiple times
         // if the kernel has conditional feedback. In that case,
         // we have to be careful about deleting the `argData` and
         // only do so after the last invocation of that function.
-
-        // Look and see if this is a kernel with conditional feedback
-        bool hasQubitMeasurementFeedback =
-            unwrap(kernelMod)
-                .lookupSymbol<mlir::func::FuncOp>("__nvqpp__mlirgen__" +
-                                                  kernelName)
-                ->hasAttrOfType<mlir::BoolAttr>("qubitMeasurementFeedback");
+        // Hence, pass it as a unique_ptr for the functor to manage its
+        // lifetime.
+        std::unique_ptr<OpaqueArguments> argData(
+            toOpaqueArgs(args, kernelMod, kernelName));
 
         // Should only have C++ going on here, safe to release the GIL
         py::gil_scoped_release release;
         return cudaq::details::runSamplingAsync(
-            [argData, kernelName, kernelMod, shots, hasQubitMeasurementFeedback,
-             &kernel]() mutable {
-              static std::size_t localShots = 0;
-              pyAltLaunchKernel(kernelName, kernelMod, *argData, {});
-              // delete the raw arg data pointer.
-              if (hasQubitMeasurementFeedback) {
-                if (localShots == shots - 1)
-                  delete argData;
-                localShots++;
-                kernel.dec_ref();
-              }
-            },
+            // Notes:
+            // (1) no Python data access is allowed in this lambda body.
+            // (2) This lambda might be executed multiple times, e.g, when the
+            // kernel contains measurement feedback.
+            cudaq::detail::make_copyable_function(
+                [argData = std::move(argData), kernel = std::move(kernelPtr),
+                 kernelName, kernelMod]() mutable {
+                  pyAltLaunchKernel(kernelName, kernelMod, *argData, {});
+                }),
```

Comment:
I don't understand how this can work. With `argData` and `kernel` being `std::unique_ptr`s, this lambda should not be copyable.

Reply:
It feels like a misleading behavior. We don't think we need to here, but it would be good to revisit this function, at least change its name to something that clearly signals that different copies of the lambda are not being created.

Reply:
Agree! I added https://github.com/NVIDIA/cuda-quantum/issues/2814 to track this discussion.